### PR TITLE
ACP-77: Write subnet public key diffs to state

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

Factored out of #3388.

After ACP-77, L1 validators specify their own BLS public keys. This means that we can no longer assume that the public keys are inherited from a primary network validator.

## How this works

Previously public key diffs were only written for primary network validators.
Now when a subnet validator is added or removed, a subnet specific key diff is written.
Additionally, non-nil public keys were only held in the validator set for the primary network validators. The state now exposes the public keys on all subnets.

## How this was tested

- [X] Existing unit tests
- [X] Expanded state diff application to include testing of the new index
- [X] Expanded state diff application to include testing of the prior subnet validator key indexing lookup

### CODE REVIEW

File: vms/platformvm/validators/manager.go [hunk 1]
```diff
@@ -271,7 +272,7 @@ func (m *manager) makePrimaryNetworkValidatorSet(
 		validatorSet,
 		currentHeight,
 		lastDiffHeight,
-		constants.PlatformChainID,
```
Comment (StephenButtolph) [inline vms/platformvm/validators/manager.go:274]: This was the wrong constant before - but they happen to have the same value... So not a bug

File: vms/platformvm/state/state_test.go [hunk 2]
```diff
@@ -109,620 +110,336 @@ func TestStateSyncGenesis(t *testing.T) {
...
-		checkStakerInState func(*require.Assertions, *state, *Staker)
-
-		// Check whether validators are duly reported in the validator set,
-		// with the right weight and showing the BLS key
-		checkValidatorsSet func(*require.Assertions, *state, *Staker)
+	const (
+		primaryValidatorDuration = 28 * 24 * time.Hour
+		primaryDelegatorDuration = 14 * 24 * time.Hour
+		subnetValidatorDuration  = 21 * 24 * time.Hour
+		subnetDelegatorDuration  = 14 * 24 * time.Hour
```
Comment (StephenButtolph) [inline vms/platformvm/state/state_test.go:120]: these are unused now

File: vms/platformvm/state/state_test.go [hunk 3]
```diff
@@ -109,620 +110,390 @@ func TestStateSyncGenesis(t *testing.T) {
...
+			expectedValidatorSetOutput: &validators.GetValidatorOutput{
+				NodeID:    primaryNetworkCurrentValidatorStaker.NodeID,
+				PublicKey: primaryNetworkCurrentValidatorStaker.PublicKey,
+				Weight:    primaryNetworkCurrentValidatorStaker.Weight,
+			},
+			expectedWeightDiff: &ValidatorWeightDiff{
+				Decrease: false,
+				Amount:   primaryNetworkCurrentValidatorStaker.Weight,
 			},
+			expectedPublicKeyDiff: maybe.Some[*bls.PublicKey](nil),
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply vms/platformvm/state/state_test.go:240]: This is not the empty value

File: vms/platformvm/state/state_test.go [hunk 4]
```diff
@@ -996,35 +767,43 @@ func TestStateAddRemoveValidator(t *testing.T) {
 	state := newTestState(t, memdb.New())
 
 	var (
-		numNodes  = 3
-		subnetID  = ids.GenerateTestID()
-		startTime = time.Now()
-		endTime   = startTime.Add(24 * time.Hour)
-		stakers   = make([]Staker, numNodes)
+		numNodes       = 5
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply vms/platformvm/state/state_test.go:770]: This is testing higher level behavior than `TestPersistStakers` was

**Comment by StephenButtolph** [comment]
@marun I refactored unrelated testing changes into a separate PR: #3487

File: vms/platformvm/state/state.go [hunk 5]
```diff
@@ -2028,164 +2042,220 @@ func (s *state) writeExpiry() error {
...
-				return err
+			weightDiff.Amount = staker.Weight
+		}
+
+		switch validatorDiff.validatorStatus {
+		case added:
+			if pk != nil {
+				// Record that the public key for the validator is being added.
+				// This means the prior value for the public key was nil.
+				err := s.validatorPublicKeyDiffsDB.Put(
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply vms/platformvm/state/state.go:2152]: We are writing the state-diff that occurred at this height. A diff could be either an Additions or a Removal. Either way the value on disk is changing, so there is information to `Put` on disk

**Review by ceyonur — APPROVED** [review]

**Pull request merged**